### PR TITLE
564282 API revision: question extensibility

### PR DESCRIPTION
**Pull request by eparovyshnaya**
The decision has been taken to close extensibility for critical access cycle parts and 'seal' final services assembly for security aims. Thus, there:
 
  - `registry` implementation is altered to have _final_ readonly registry class
  - extensibility of requirement resolution services registry is closed: readonly registry is used
  - both requirement resolution services we have now (`BundleRequirements` and `ComponentRequirements`) are detached from OSGi to be accessible for sealing
  - contract tests are designed for `Framework` and resuirement resolution service (new `lic.api.tests` bundle)
  - `seal.demo` bundle is added to supply final assembly of passage runtime (packed instance of `Framework` interface) for out-of-the-box facilities demontration purposes
  - `seal.demo.tests` bundle contains both contract and content tests for current `DemoFramework`
  - `demo` feature is appended

### CODE REVIEW

File: tests/org.eclipse.passage.seal.demo.tests/src/org/eclipse/passage/seal/demo/tests/DemoFrameworkContractTest.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package org.eclipse.passage.seal.demo.tests;
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.seal.demo.tests/src/org/eclipse/passage/seal/demo/tests/DemoFrameworkContractTest.java:1]: header

File: tests/org.eclipse.passage.seal.demo.tests/src/org/eclipse/passage/seal/demo/tests/DemoFrameworkContractTest.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import org.eclipse.passage.lic.internal.api.Framework;
+
+import java.util.Optional;
+
+import org.eclipse.passage.lic.api.tests.FrameworkContractTest;
+import org.eclipse.passage.lic.internal.api.registry.Registry;
+import org.eclipse.passage.seal.internal.demo.DemoFrameworkSupplier;
+
+@SuppressWarnings("restriction")
+public final class DemoFrameworkContractTest extends FrameworkContractTest{
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.seal.demo.tests/src/org/eclipse/passage/seal/demo/tests/DemoFrameworkContractTest.java:12]: format

File: tests/org.eclipse.passage.seal.demo.tests/src/org/eclipse/passage/seal/demo/tests/DemoFrameworkContentTest.java [hunk 3]
```diff
@@ -1,16 +1,4 @@
...
- * terms of the Eclipse Public License 2.0 which is available at
- * https://www.eclipse.org/legal/epl-2.0/.
- *
- * SPDX-License-Identifier: EPL-2.0
- *
- * Contributors:
- *     ArSysOp - initial API and implementation
- *******************************************************************************/
-package org.eclipse.passage.lic.internal.equinox;
+package org.eclipse.passage.seal.demo.tests;
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.seal.demo.tests/src/org/eclipse/passage/seal/demo/tests/DemoFrameworkContentTest.java:1]: header

File: tests/org.eclipse.passage.seal.demo.tests/META-INF/MANIFEST.MF [hunk 4]
```diff
@@ -0,0 +1,16 @@
+Manifest-Version: 1.0
+Automatic-Module-Name: org.eclipse.passage.seal.demo.tests
+Bundle-ManifestVersion: 2
+Bundle-SymbolicName: org.eclipse.passage.seal.demo.tests
+Bundle-Version: 0.5.0.qualifier
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.seal.demo.tests/META-INF/MANIFEST.MF:5]: why 0.5?

File: tests/org.eclipse.passage.seal.demo.tests/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+Bundle-Name: %Bundle-Name
+Bundle-Vendor: %Bundle-Vendor
+Bundle-Copyright: %Bundle-Copyright
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Require-Bundle: org.junit;bundle-version="4.12.0",
+ org.eclipse.osgi.services;bundle-version="0.0.0",
+ org.eclipse.passage.seal.demo;bundle-version="0.0.0",
+ org.eclipse.passage.lic.api.tests;bundle-version="0.0.0",
+ org.eclipse.passage.lic.api,
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.seal.demo.tests/META-INF/MANIFEST.MF:14]: I assume this requires API 1.0

File: tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeResolvedRequirements.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.util.Collection;
+import java.util.Collections;
+
+import org.eclipse.passage.lic.internal.api.registry.StringServiceId;
+import org.eclipse.passage.lic.internal.api.requirements.Requirement;
+import org.eclipse.passage.lic.internal.api.requirements.ResolvedRequirements;
+
+@SuppressWarnings("restriction")
+public final class FakeResolvedRequirements implements ResolvedRequirements{
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeResolvedRequirements.java:11]: format

File: tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeResolvedRequirements.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package org.eclipse.passage.lic.api.tests.fakes;
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeResolvedRequirements.java:1]: header

File: tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeRequirement.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
+package org.eclipse.passage.lic.api.tests.fakes;
+
+import org.eclipse.passage.lic.internal.api.requirements.Feature;
+import org.eclipse.passage.lic.internal.api.requirements.Requirement;
+import org.eclipse.passage.lic.internal.api.restrictions.RestrictionLevel;
+
+@SuppressWarnings("restriction")
+public final class FakeRequirement implements Requirement{
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeRequirement.java:8]: format

File: tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeRequirement.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
+package org.eclipse.passage.lic.api.tests.fakes;
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeRequirement.java:1]: header

File: tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeFeature.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
+package org.eclipse.passage.lic.api.tests.fakes;
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/fakes/FakeFeature.java:1]: header

File: tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/ResolvedReqiorementsContractTest.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
+package org.eclipse.passage.lic.api.tests;
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/ResolvedReqiorementsContractTest.java:1]: header

File: tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/FrameworkContractTest.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
+package org.eclipse.passage.lic.api.tests;
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/FrameworkContractTest.java:1]: header

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/internal/api/FrameworkSupplier.java [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package org.eclipse.passage.lic.internal.api;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/internal/api/FrameworkSupplier.java:1]: header

File: bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/requirements/ComponentRequirements.java [hunk 14]
```diff
@@ -40,41 +37,27 @@
...
  */
 @SuppressWarnings("restriction")
-@Component
 public final class ComponentRequirements implements ResolvedRequirements {
 
 	private final Logger logger = LoggerFactory.getLogger(BundleRequirements.class);
 
-	private Optional<BundleContext> context = Optional.empty();
-	private Optional<ServiceComponentRuntime> runtime = Optional.empty();
+	private Optional<BundleContext> context;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/requirements/ComponentRequirements.java:44]: final

File: tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/TemporaryTest.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
+package org.eclipse.passage.lic.api.tests;
```
Comment (ruspl-afed) [inline tests/org.eclipse.passage.lic.api.tests/src/org/eclipse/passage/lic/api/tests/TemporaryTest.java:1]: header

File: features/org.eclipse.passage.demo.feature/feature.xml [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+	terms of the Eclipse Public License 2.0 which is available at
+	https://www.eclipse.org/legal/epl-2.0/.
+
+	SPDX-License-Identifier: EPL-2.0
+
+	Contributors:
+		ArSysOp - initial API and implementation
+-->
+<feature
+      id="org.eclipse.passage.loc.workbench.feature"
```
Comment (ruspl-afed) [inline features/org.eclipse.passage.demo.feature/feature.xml:15]: wrong feature id

File: features/org.eclipse.passage.demo.feature/feature.xml [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+      label="%featureName"
+      version="0.6.200.qualifier"
```
Comment (ruspl-afed) [inline features/org.eclipse.passage.demo.feature/feature.xml:17]: wrong feature version

File: features/org.eclipse.passage.demo.feature/feature.properties [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#     ArSysOp - initial API and implementation
+###############################################################################
+
+featureName=Passage Demo
+providerName=Eclipse Passage
+description=Passage Licensing Operator Components: Workbench
```
Comment (ruspl-afed) [inline features/org.eclipse.passage.demo.feature/feature.properties:17]: wrong feature description

File: features/org.eclipse.passage.demo.feature/.project [hunk 19]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>org.eclipse.passage.loc.workbench.feature</name>
```
Comment (ruspl-afed) [inline features/org.eclipse.passage.demo.feature/.project:3]: wrong project name

File: bundles/org.eclipse.passage.seal.demo/src/org/eclipse/passage/seal/internal/demo/DemoFramework.java [hunk 20]
```diff
@@ -0,0 +1,33 @@
+package org.eclipse.passage.seal.internal.demo;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.seal.demo/src/org/eclipse/passage/seal/internal/demo/DemoFramework.java:1]: header

File: bundles/org.eclipse.passage.seal.demo/META-INF/MANIFEST.MF [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+Automatic-Module-Name: org.eclipse.passage.seal.demo
+Bundle-ManifestVersion: 2
+Bundle-SymbolicName: org.eclipse.passage.seal.demo
+Bundle-Version: 0.1.0.qualifier
+Bundle-Name: %Bundle-Name
+Bundle-Vendor: %Bundle-Vendor
+Bundle-Copyright: %Bundle-Copyright
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Require-Bundle: org.eclipse.osgi.services;bundle-version="3.8.0",
+ org.eclipse.passage.lic.api,
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.seal.demo/META-INF/MANIFEST.MF:11]: I would specify version requirements

File: bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/requirements/ComponentRequirements.java [hunk 22]
```diff
@@ -40,41 +37,27 @@
...
+	private Optional<ServiceComponentRuntime> runtime;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/requirements/ComponentRequirements.java:45]: final

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/registry/ReadOnlyRegistry.java [hunk 23]
```diff
@@ -0,0 +1,15 @@
+package org.eclipse.passage.lic.internal.base.registry;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/registry/ReadOnlyRegistry.java:1]: header

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/registry/BaseRegistry.java [hunk 24]
```diff
@@ -23,31 +24,37 @@
...
+ * </p>
+ * 
+ * @param <S> type of {@linkplain Service} to keep
  */
 @SuppressWarnings("restriction")
-public final class BaseRegistry<I extends ServiceId, S extends Service<I>> implements Registry<I, S> {
+public abstract class BaseRegistry<I extends ServiceId, S extends Service<I>> implements Registry<I, S> {
 
-	private final Supplier<Collection<S>> services;
+	protected final Map<I, S> services;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/registry/BaseRegistry.java:43]: What is the reason to have it protected?

File: tests/org.eclipse.passage.seal.demo.tests/META-INF/MANIFEST.MF (hunk 4 above)
Comment (eparovyshnaya) [reply tests/org.eclipse.passage.seal.demo.tests/META-INF/MANIFEST.MF:5]: Oh, dear

File: tests/org.eclipse.passage.seal.demo.tests/META-INF/MANIFEST.MF (hunk 5 above)
Comment (eparovyshnaya) [reply tests/org.eclipse.passage.seal.demo.tests/META-INF/MANIFEST.MF:14]: It only uses _internal_ parts not yet pushed to the official api level. Current `lic.api` is of version _0.7.0_. Can you please explain your expectation in some details?

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/registry/BaseRegistry.java (hunk 24 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/registry/BaseRegistry.java:43]: `RuntimeRegistry` alters this map (registers and unregisters service)

File: bundles/org.eclipse.passage.seal.demo/META-INF/MANIFEST.MF [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+Bundle-Vendor: %Bundle-Vendor
+Bundle-Copyright: %Bundle-Copyright
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Require-Bundle: org.eclipse.osgi.services;bundle-version="3.8.0",
+ org.eclipse.passage.lic.api;bundle-version="0.7.0",
+ org.eclipse.passage.lic.base;bundle-version="0.6.100",
+ org.eclipse.passage.lic.equinox;bundle-version="0.6.0"
+Service-Component: OSGI-INF/*.xml
+Bundle-ActivationPolicy: lazy
+Export-Package: org.eclipse.passage.seal.internal.demo
```
[the remark this answers is not in the document]
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.seal.demo/META-INF/MANIFEST.MF]: Done

**Pull request merged**